h/services/user_signup.py

```diff
@@ -80,7 +82,14 @@ def signup(self, require_activation=True, **kwargs):
 
         # Create a new activation for the user
         if require_activation:
-            self._require_activation(user)
+            try:
+                self._require_activation(user)
+            except IntegrityError:
```

Comment:
Is it safe to rely on the exception type from psycopg2 here? Since that is the layer below SQLAlchemy, which is what our code interacts with. I notice in the user_update.py service we catch a SQLAlchemy exception instead: https://github.com/hypothesis/h/blob/fe8ab1cab3131be4be464e47c59dc8bfe46842d2/h/services/user_update.py#L47. The fact that we're parsing a string exception message in that code is not great, I wonder if there are better options?

Reply:
sqlalchemy relies on the underlying db library and makes calls out to it but maybe it wraps the error message? I'll look into it.

Reply:
Using the SQLAlchemy layer has been the pattern the rest of `h` uses, so consistency seems like a good idea...

Reply:
So I've been reading up on this. sqlalchemy exceptions have an "orig" attribute where we can get the error code and the error message:
```
>> e.orig.args
('duplicate key value violates unique constraint "uq__user__email"\nDETAIL:  Key (email, authority)=(pony@localhost.com, localhost) already exists.\n',)
>> e.orig.pgcode
'23505'
>> e.orig.pgerror
'ERROR:  duplicate key value violates unique constraint "uq__user__email"\nDETAIL:  Key (email, authority)=(pony@localhost.com, localhost) already exists.\n'
```
That being said, we can get at this from the original sqlalchemy error but note sqlalchemy also provides sqlalchemy.exc.IntegrityError which seems like what we should use here. I'm not sure why we aren't catching more specific exceptions in other places in the application but it seems like we should be doing that?